### PR TITLE
Revert lowering of wrapping_* intrinsics

### LINKED ISSUE #82646: Wrong error on wrapping arithmetic with -Zmir-opt-level=2
I tried this code with `-Z mir-opt-level=2`:

```rust
fn main() {
    println!("{}", 1_u32.wrapping_sub(2_u32));
}
```

It compiles and runs without mir-opt, but with `-Z mir-opt-level=2` or above I get:

```
error: this arithmetic operation will overflow
   --> src/main.rs:2:13
    |
2   |       println!("{}", 1_u32.wrapping_sub(2_u32));
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ attempt to compute `1_u32 - 2_u32`, which would overflow
    |
```

### Meta

`rustc --version --verbose`:
```
rustc 1.52.0-nightly (0148b971c 2021-02-18)
binary: rustc
commit-hash: 0148b971c921a0831fbf3357e5936eec724e3566
commit-date: 2021-02-18
host: x86_64-apple-darwin
release: 1.52.0-nightly
LLVM version: 11.0.1
```

### PR DESCRIPTION
**Pull request by tmiasko**
The const propagation needs to distinguish wrapping intrinsics from binary operations for linting purposes.

Closes #82646.

### CODE REVIEW

**Comment by tmiasko** [comment]
cc @RalfJung, @bjorn3

File: src/test/ui/const_prop/inline_spans.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 fn main() {
     let _ = add(u8::MAX, 1);
     //~^ ERROR this arithmetic operation will overflow
+    let _ok = 1_u32.wrapping_sub(2_u32);
```
Comment (RalfJung) [inline src/test/ui/const_prop/inline_spans.rs:9]: Maybe add a comment saying that this should *not* lint.

**Comment by RalfJung** [comment]
Ah, that's a shame... another bullet point on the list of reasons why it is bad to do linting as part of an optimization (in particular an optimization that runs so late in the pipeline).

**Comment by RalfJung** [comment]
Cc @oli-obk

**Comment by oli-obk** [comment]
We could "just" stop linting on arithmetic and instead only lint on the assert terminators. This means that lints change with debug assertions, but maybe that's not problematic? This can already happen with `#[cfg(debug_assertions)]`.

I would prefer to find a fix over this PR. If that fix is to duplicate the const prop pass into a basic one for linting to run early and one for actual propagation, then we should do that.

This is actually fairly easy to do, we already have a mode for only linting and not optimizing, so we can run the const prop twice, once with linting but without optimizations, and once with optimizations but no linting. Does that sound reasonable as a first step?

**Comment by RalfJung** [comment]
>  This means that lints change with debug assertions

I'd rather not do that, it's a testing nightmare and a bad user experience.

**Comment by RalfJung** [comment]
> If that fix is to duplicate the const prop pass into a basic one for linting to run early and one for actual propagation, then we should do that.

That sounds reasonable to me... the lint version doesn't need to "propagate" much, e.g. I think it's fine not to warn in `1/ (1-1)`.

> This is actually fairly easy to do, we already have a mode for only linting and not optimizing, so we can run the const prop twice, once with linting but without optimizations, and once with optimizations but no linting. Does that sound reasonable as a first step?

I guess that also works. It means the lint becomes more expensive when const-prop is made more fancy, but maybe that's unlikely or okay.

**Comment by oli-obk** [comment]
> I guess that also works. It means the lint becomes more expensive when const-prop is made more fancy, but maybe that's unlikely or okay.

I do mention this as a first step, because it is the fastest way forward. There are so many problems around const prop lints that are entirely unrelated to const prop opts, moving to a much simpler version as you suggested sound like a good plan. We can even start out by literally copying the const propagator and slowly stripping the copy that is just for linting.

**Comment by tmiasko** [comment]
Separating the lint sounds like a good idea.

The only part I would be worried about is potential performance impact, I would but the constant propagation cost as roughly all other remaining transformations from optimization phase combined, although that could be misleading impression left by a query system.

**Pull request closed without merging**